### PR TITLE
repo: leverage stored index to calculate shortest prefix in commit id space

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: lib/src/index.rs [hunk 1]
```diff
@@ -1955,6 +2004,91 @@ mod tests {
...
+
+        let id_3 = CommitId::from_hex("055444");
+        let id_4 = CommitId::from_hex("055555");
+        let id_5 = CommitId::from_hex("033333");
+        index.add_commit_data(id_3.clone(), new_change_id(), &[]);
+        index.add_commit_data(id_4.clone(), new_change_id(), &[]);
+        index.add_commit_data(id_5.clone(), new_change_id(), &[]);
+
+        // Local lookup in readonly index, commit_id exists.
+        let readonly_index = index.segment_parent_file().unwrap().as_ref();
```
Comment (martinvonz) [inline lib/src/index.rs:2089]: This is the same as `initial_file`, right? I think it's clearer to use that (and pass in a clone into `MutableIndex::incremental()` above).

File: lib/src/index.rs [hunk 2]
```diff
@@ -698,6 +698,18 @@ trait IndexSegment {
...
+        let parent_file = self.0.segment_parent_file();
+        iter::successors(parent_file, |file| file.segment_parent_file())
+    }
+
+    fn ancestor_index_segments(&self) -> impl Iterator<Item = &dyn IndexSegment> {
+        iter::once(self.0).chain(
+            self.ancestor_files_without_local()
+                .map(|file| file.as_ref() as &dyn IndexSegment),
+        )
+    }
```
Comment (martinvonz) [inline lib/src/index.rs:722]: Neat tricks using the `iter` module. Thanks for teaching my about those.

**Review by martinvonz — APPROVED** [review]
Thanks!

File: lib/src/index.rs (hunk 1 above)
Comment (yuja) [reply lib/src/index.rs]: Indeed. Fixed, thanks.

**Pull request merged**